### PR TITLE
Replace Vegeta with custom HTTP executor (ASAP model)

### LINKED ISSUE #23: Replace Vegeta with custom HTTP executor (ASAP model)
## Context

kube-booster currently uses Vegeta (HTTP load testing library) to send warmup requests. Vegeta spreads requests evenly over a configured duration (rate-based "open model"), which is misaligned with warmup requirements: we need back-to-back requests that complete as fast as possible.

This is **Phase 1** of the warmup engine replacement. See also:
- Phase 2: #24 — Cross-pod concurrency control
- Phase 3: #25 — gRPC support
- Phase 4: #26 — Scenario-based execution with CRD

## What Changes

- `Config.Duration` becomes a **maximum timeout** instead of a spread-across duration
- Requests fire **back-to-back** (closed model / ASAP) instead of rate-spread (open model)
- Vegeta dependency (`github.com/tsenart/vegeta/v12`) is removed entirely

### Semantic Change

| | Current (Vegeta) | New (ASAP) |
|---|---|---|
| 3 requests, Duration=30s | 1 request every 10s → finishes in ~30s | Back-to-back → finishes in ~300ms |
| Duration meaning | Spread-across duration | Maximum timeout |
| Model | Open (fixed rate) | Closed (feedback loop) |

## Files to Change

| Action | File | What |
|--------|------|------|
| Modify | `pkg/warmup/executor.go` | Remove `VegetaExecutor`, `NewVegetaExecutor`, vegeta import. Keep `Executor` interface, `ErrNoPodIP`, `WarmupError` |
| Create | `pkg/warmup/http_executor.go` | New `HTTPExecutor` with back-to-back request loop using `net/http` |
| Delete | `pkg/warmup/executor_test.go` | Replaced by new test file |
| Create | `pkg/warmup/http_executor_test.go` | Tests for HTTPExecutor (table-driven, httptest.Server) |
| Modify | `cmd/controller/main.go` | Line 123: `NewVegetaExecutor` → `NewHTTPExecutor` |
| Modify | `pkg/controller/pod_controller.go` | Line 128: Update comment (remove "Vegeta" reference) |
| Modify | `go.mod` | Remove `github.com/tsenart/vegeta/v12`; run `go mod tidy` |

### `HTTPExecutor` (`pkg/warmup/http_executor.go`)

```go
type HTTPExecutor struct {
    logger logr.Logger
    client *http.Client
}

func NewHTTPExecutor(logger logr.Logger) *HTTPExecutor

func (e *HTTPExecutor) Execute(ctx context.Context, config *Config) *Result
```

Core loop:
```go
warmupCtx, cancel := context.WithTimeout(ctx, config.Duration)
defer cancel()
for i := 0; i < config.RequestCount; i++ {
    if warmupCtx.Err() != nil { break }
    // Create request with warmup headers
    // Send request, record latency
    // Close response body immediately
}
```

- HTTP requests include headers: `User-Agent: kube-booster/1.0`, `X-Warmup-Request: true` (unchanged)
- Success criteria: `statusCode >= 200 && statusCode < 400`
- Latency percentiles via sorted-slice approach (adequate for N=3..100):

```go
func calculatePercentiles(latencies []time.Duration) (p50, p99 time.Duration)
```

### Error Handling

- Context cancelled before any request → `result.Error = ctx.Err()`, `result.Success = false`
- All requests fail with network errors → `result.Success = false`, `RequestsFailed = RequestCount`
- Duration timeout hit mid-loop → loop breaks, partial results returned
- Empty PodIP → returns immediately with `ErrNoPodIP` (unchanged behavior)

## Tests

- `TestHTTPExecutor_Execute` — table-driven: successful warmup, server errors, no pod IP
- `TestHTTPExecutor_Execute_ContextCancellation` — slow server + short timeout
- `TestHTTPExecutor_Execute_MetricsCollection` — verify latency/duration fields populated
- `TestCalculatePercentiles` — unit test: empty slice, single element, known values

## Verification

```bash
make test          # all tests pass
make lint          # no vegeta imports remain
make build         # binary builds
go mod tidy        # vegeta removed from go.mod/go.sum
```

## Documentation Updates

- `docs/USAGE.md`: Update FAQ on warmup request distribution to explain ASAP model
- `docs/IMPLEMENTATION_SUMMARY.md`: Update executor section
- `docs/DEVELOPMENT.md`: Update warmup package section

### PR DESCRIPTION
**Pull request by hhiroshell**
## Summary

Implements #23

- Replace Vegeta-based warmup executor with custom `HTTPExecutor` using `net/http`
- Requests now fire back-to-back (ASAP/closed model) instead of rate-spread (open model)
- Rename `Duration` → `Timeout` across annotations, constants, and config per owner feedback
- Remove `github.com/tsenart/vegeta/v12` dependency entirely

### Code
- **`pkg/warmup/http_executor.go`** (new): `HTTPExecutor` with back-to-back request loop, `calculatePercentiles` helper, `Executor` interface, `ErrNoPodIP`, `WarmupError`
- **`pkg/warmup/http_executor_test.go`** (new): Table-driven tests, context cancellation, metrics collection, ASAP behavior validation, percentile calculation tests
- **`pkg/warmup/executor.go`** (deleted): Removed `VegetaExecutor` and vegeta import
- **`pkg/warmup/executor_test.go`** (deleted): Replaced by new test file
- **`pkg/webhook/constants.go`**: `AnnotationWarmupDuration` → `AnnotationWarmupTimeout`, annotation key `warmup-duration` → `warmup-timeout`
- **`pkg/warmup/config.go`**: `DefaultDuration` → `DefaultTimeout`, `Config.Duration` → `Config.Timeout`
- **`pkg/warmup/config_test.go`**: Updated all field refs and annotation keys
- **`cmd/controller/main.go`**: `NewVegetaExecutor` → `NewHTTPExecutor`
- **`pkg/controller/pod_controller.go`**: Updated comment and `config.Duration` → `config.Timeout`
- **`go.mod`/`go.sum`**: Removed vegeta and transitive deps (tdigest, dnscache)

### Documentation
- **CLAUDE.md, README.md, docs/USAGE.md, docs/DEVELOPMENT.md, docs/IMPLEMENTATION_SUMMARY.md**: Updated annotation names, executor descriptions, request model explanation
- **config/samples/sample_deployment.yaml**: Updated annotation key
- **hack/quick_test.sh**: Updated annotation key in smoke test

## Breaking Changes

- Annotation renamed: `kube-booster.io/warmup-duration` → `kube-booster.io/warmup-timeout`
- Warmup behavior: requests complete much faster (ASAP vs rate-spread)

## Testing

- `make test` — all tests pass (89.2% warmup coverage)
- `make build` — binary builds successfully
- `go mod tidy` — vegeta completely removed from go.mod/go.sum
- No remaining references to `vegeta`, `warmup-duration`, or `DefaultDuration` in codebase

Closes #23

---
*PR generated with [Claude Code](https://claude.ai/code) - github-devflow:implement*

### CODE REVIEW

File: pkg/warmup/http_executor.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		req.Header.Set("X-Warmup-Request", "true")
+
+		resp, err := e.client.Do(req)
+		latency := time.Since(reqStart)
+
+		if err != nil {
+			failCount++
+			// If context was cancelled/timed out, record the error and break
+			if warmupCtx.Err() != nil {
+				result.Error = ctx.Err()
```
Comment (hhiroshell) [inline pkg/warmup/http_executor.go]: **[Logic]** :rotating_light: When a request fails because `warmupCtx` expired, `result.Error = ctx.Err()` reads from the **parent** context instead of `warmupCtx.Err()`. If only the inner timeout (`config.Timeout`) fired, the parent context is still alive and `ctx.Err()` returns `nil`, silently swallowing the timeout error.

**[Design]** :rotating_light: Same issue — the wrong context is used to report the error. All three sites (line 85, 110, 114) should use `warmupCtx.Err()` instead of `ctx.Err()`.

**[Security]** :information_source: This suppresses timeout error reporting in the result and audit trail, reducing observability of misbehaving warmup configurations.

**Fix:** Change to `result.Error = warmupCtx.Err()`

File: pkg/warmup/http_executor.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		} else {
+			failCount++
+		}
+	}
+
+	totalDuration := time.Since(start)
+
+	// If no requests were sent and context was cancelled, set the error
+	if successCount == 0 && failCount == 0 && warmupCtx.Err() != nil {
+		result.Error = ctx.Err()
```
Comment (hhiroshell) [inline pkg/warmup/http_executor.go]: **[Logic]** :rotating_light: Same `ctx` vs `warmupCtx` confusion as line 85. The guard correctly checks `warmupCtx.Err() != nil`, but then assigns `result.Error = ctx.Err()` (the outer context). When the inner `warmupCtx` times out but the outer `ctx` is still alive, `ctx.Err()` returns `nil`.

**Fix:** Change lines 110 and 114 to use `warmupCtx.Err()` instead of `ctx.Err()`.

File: pkg/warmup/http_executor.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+			continue
+		}
+		if _, err := io.Copy(io.Discard, resp.Body); err != nil {
+			e.logger.V(2).Info("failed to drain response body", "error", err)
+		}
+		if err := resp.Body.Close(); err != nil {
+			e.logger.V(2).Info("failed to close response body", "error", err)
+		}
+
+		latencies = append(latencies, latency)
```
Comment (hhiroshell) [inline pkg/warmup/http_executor.go]: **[Logic]** :warning: Latency is appended to `latencies` (line 97) unconditionally after a successful HTTP round-trip, but then the status code is checked (lines 99-103) and `failCount` is incremented for non-2xx/3xx responses. This means latencies from 4xx/5xx responses are included in the percentile calculation even though those requests are counted as failures.

**[Design]** :warning: This conflates transport latency with application-level success, making P50/P99 metrics potentially misleading when failures occur. Consider recording latency only for successful responses, or document what population the percentiles represent.

**[Performance]** :information_source: The pre-allocated latencies slice (capacity = `config.RequestCount`) may not match actual usage when many requests fail.

File: pkg/warmup/config.go [hunk 4]
```diff
@@ -77,16 +77,16 @@ func ParseConfig(pod *corev1.Pod) (*Config, error) {
...
-				return config, fmt.Errorf("invalid warmup-duration value %q: %w", durationStr, err)
+				return config, fmt.Errorf("invalid warmup-timeout value %q: %w", timeoutStr, err)
 			}
-			if duration < time.Second {
-				return config, fmt.Errorf("warmup-duration must be at least 1s, got %v", duration)
+			if timeout < time.Second {
+				return config, fmt.Errorf("warmup-timeout must be at least 1s, got %v", timeout)
 			}
-			config.Duration = duration
+			config.Timeout = timeout
```
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/config.go]: Please set the maximum timeout to 5 minutes. I think we should start by prioritizing the safe operation of kube-booster.
Let's revisit this once we have requirements from users and operational knowledge from real-world environments.

File: pkg/warmup/http_executor.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	logger logr.Logger
+	client *http.Client
+}
+
+// NewHTTPExecutor creates a new HTTPExecutor
+func NewHTTPExecutor(logger logr.Logger) *HTTPExecutor {
+	return &HTTPExecutor{
+		logger: logger,
+		client: &http.Client{
+			Timeout: 10 * time.Second,
```
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/http_executor.go]: Dynamically adjusting the per-request HTTP client timeout may lead to unpredictable behavior for users. We already have a guardrail in place to protect the kube-booster Pod by setting the warm-up timeout, so I don’t think we need to address this.

TLS support should be out of scope for this PR.

File: pkg/warmup/config.go [hunk 6]
```diff
@@ -77,16 +77,16 @@ func ParseConfig(pod *corev1.Pod) (*Config, error) {
 			config.RequestCount = reqCount
```
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/config.go]: The maximum should be set to around 12,000. This is because the threshold for JVM level 4 (C2) JIT compilation occurs after the same code has been executed roughly 10,000 times.

File: pkg/controller/pod_controller.go [hunk 7]
```diff
@@ -124,9 +124,9 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 	config.PodName = pod.Name
 	config.PodNamespace = pod.Namespace
 
-	// Execute warmup with context timeout slightly longer than duration
-	// to allow Vegeta to finish cleanly after the attack completes
-	contextTimeout := config.Duration + 5*time.Second
+	// Execute warmup with context timeout slightly longer than the configured timeout
+	// to allow the executor to finish cleanly
+	contextTimeout := config.Timeout + 5*time.Second
```
Comment (hhiroshell) [inline pkg/controller/pod_controller.go]: Does this become somewhat meaningless after we change the duration to a timeout? The new HTTP executor does not always use the full timeout duration.

File: pkg/warmup/http_executor.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+		if resp.StatusCode >= 200 && resp.StatusCode < 400 {
+			successCount++
+		} else {
+			failCount++
+		}
+	}
+
+	totalDuration := time.Since(start)
+
+	// If no requests were sent and context was cancelled, set the error
```
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/http_executor.go]: To `github-devflow:reply`

Please clarify what the “two separate early-return paths” refer to before commenting on this thread. It will help all reviewers and implementers better understand the point.

File: pkg/warmup/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/config.go]: Please don't forget to update the documentation related to the annotation.

File: docs/IMPLEMENTATION_SUMMARY.md [hunk 9]
```diff
@@ -121,7 +121,7 @@ k8s.io/apimachinery v0.35.0
 k8s.io/client-go v0.35.0
 sigs.k8s.io/controller-runtime v0.23.0
 go.uber.org/zap v1.27.0
-github.com/tsenart/vegeta/v12 v12.12.0
+(vegeta removed — using net/http directly)
```
Comment (hhiroshell) [inline docs/IMPLEMENTATION_SUMMARY.md]: No need to mention about removal of Vegeta

File: pkg/warmup/http_executor.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+// calculatePercentiles computes P50 and P99 from a slice of latencies.
+func calculatePercentiles(latencies []time.Duration) (p50, p99 time.Duration) {
+	n := len(latencies)
+	if n == 0 {
+		return 0, 0
+	}
+	sort.Slice(latencies, func(i, j int) bool { return latencies[i] < latencies[j] })
+	p50 = latencies[n*50/100]
```
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/http_executor.go]: Fixed in commit 44d3a93.

Switched from the floor formula (`n*p/100`) to the nearest-rank method (`max(0, ceil(n*p/100)-1)`). This correctly picks the median for small sample sizes (e.g., for n=2 it now returns index 0 instead of 1). Updated test expectations accordingly.

---
*Reply generated with [Claude Code](https://claude.ai/code) - github-devflow:fix*

File: pkg/warmup/http_executor.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+	return result
+}
+
+// calculatePercentiles computes P50 and P99 from a slice of latencies.
+func calculatePercentiles(latencies []time.Duration) (p50, p99 time.Duration) {
+	n := len(latencies)
+	if n == 0 {
+		return 0, 0
+	}
+	sort.Slice(latencies, func(i, j int) bool { return latencies[i] < latencies[j] })
```
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/http_executor.go]: No action taken: with the proposed max of 12,000 requests, `sort.Slice` completes in microseconds. The warmup HTTP round-trips dominate by orders of magnitude. Keeping the current O(n log n) approach for simplicity.

---
*Reply generated with [Claude Code](https://claude.ai/code) - github-devflow:fix*

File: pkg/warmup/http_executor.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+	"context"
+	"io"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/go-logr/logr"
+)
+
+// Executor executes warmup requests
```
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/http_executor.go]: Created issue #28 to track this.

https://github.com/hhiroshell/kube-booster/issues/28

Will extract the `Executor` interface to `executor.go` and move error types to `errors.go` alongside the gRPC executor implementation in Phase 2.

---
*Reply generated with [Claude Code](https://claude.ai/code) - github-devflow:fix*

File: pkg/warmup/http_executor.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+	e.logger.V(1).Info("starting warmup",
+		"pod", config.PodName,
+		"namespace", config.PodNamespace,
+		"endpoint", endpoint,
+		"requestCount", config.RequestCount,
+		"timeout", config.Timeout)
+
+	// Create context with timeout
+	warmupCtx, cancel := context.WithTimeout(ctx, config.Timeout)
```
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/http_executor.go]: No action taken: the `ctx.Err()` → `warmupCtx.Err()` bug is now fixed (commit f1843cf), so the inner timeout correctly reports errors. The +5s grace period still serves as a safety net for executor cleanup, and its purpose is now documented (commit cbd6631).

---
*Reply generated with [Claude Code](https://claude.ai/code) - github-devflow:fix*

File: pkg/warmup/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hhiroshell) [reply pkg/warmup/config.go]: Fixed in commit 06f6932.

Added `MaxTimeout = 5 * time.Minute` constant and validation in `ParseConfig`. Also added a test case for exceeding the maximum and updated the annotation table in `docs/USAGE.md` to show the range (1s-5m).

---
*Reply generated with [Claude Code](https://claude.ai/code) - github-devflow:fix*

File: docs/IMPLEMENTATION_SUMMARY.md (hunk 9 above)
Comment (hhiroshell) [reply docs/IMPLEMENTATION_SUMMARY.md]: Fixed in commit cbd6631.

Removed the Vegeta removal note from the dependencies list.

---
*Reply generated with [Claude Code](https://claude.ai/code) - github-devflow:fix*

File: pkg/warmup/http_executor_test.go [hunk 14]
```diff
@@ -198,6 +196,95 @@ func TestVegetaExecutor_Execute_MetricsCollection(t *testing.T) {
 	}
 }
 
+func TestHTTPExecutor_Execute_BackToBack(t *testing.T) {
```
Comment (hhiroshell) [inline pkg/warmup/http_executor_test.go]: As we are moving to the costom executor, the back-to-back behavior is unlikely to be broken going forward. So I think we don't need to this test.
